momadm_benchmarks/envs/congestion_game/congestion_game.py

```diff
@@ -0,0 +1,406 @@
+"""Multi-Objective Congestion Game.
+
+From Ramos, G. D. O., Radulescu, R., Nowe, A., & Tavares, A. R. (2020). Toll-based learning for minimising congestion under heterogeneous preferences.
+"""
+
+import functools
+import json
+import os
+import random
+from collections import defaultdict
+from typing_extensions import override
+
+import networkx as nx
+import numpy as np
+from gymnasium.logger import warn
+from gymnasium.spaces import Box, Discrete
+from pettingzoo.utils import wrappers
+from sympy import diff, lambdify, sympify
+
+from momadm_benchmarks.utils.conversions import mo_parallel_to_aec
+from momadm_benchmarks.utils.env import MOParallelEnv
+
+
+def parallel_env(**kwargs):
+    """Env factory function for the congestion game."""
+    return raw_env(**kwargs)
+
+
+def env(**kwargs):
+    """Auto-wrapper for the congestion game.
+
+    Args:
+        **kwargs: keyword args to forward to the parallel_env function.
+
+    Returns:
+        A fully wrapped AEC env
+    """
+    env = parallel_env(**kwargs)
+    # convert parallel version of the env to an AEC version
+    env = mo_parallel_to_aec(env)
+
+    # this wrapper helps error handling for discrete action spaces
+    env = wrappers.AssertOutOfBoundsWrapper(env)
+    return env
+
+
+def raw_env(**kwargs):
+    """Env factory function for the congestion game."""
+    return MOCongestionGame(**kwargs)
+
+
+class MOCongestionGame(MOParallelEnv):
+    """Environment for MO Congestion Game problem.
+
+    The init method takes in environment arguments and should define the following attributes:
+    - possible_agents
+    - action_spaces
+    - observation_spaces
+    These attributes should not be changed after initialization.
+    """
+
+    def __init__(
+        self,
+        problem_name="Braess_1_4200_10_c1",
+        num_agents=4200,
+        toll_mode="mct",
+        random_toll_percentage=0.1,
+        num_timesteps=1,
+        render_mode=None,
+    ):
+        """Initializes the congestion game.
+
+        Args:
+            problem_name: the name of the network that will be used
+            num_agents: number of agents in the network
+            toll_mode: the tolling mode that is used, tolls are either placed randomly "random" or using marginal cost tolling "mct"
+            random_toll_percentage: in the case of random tolling the percentage of roads that will be taxed
+            num_timesteps: number of timesteps (stateless, therefore always 1 timestep)
+            render_mode: render mode
+        """
+        # the maximum length of a route in the chosen problem
+        self._max_route_length = 0
```

Comment:
I'm nitpicking here but I think it's cleaner to return the max_route_length when calling `read_problem`, making it clearer on where it is assigned